src/Microsoft.SqlTools.ServiceLayer/QueryExecution/Batch.cs

```diff
@@ -364,6 +364,8 @@ private async Task ExecuteOnce(DbConnection conn, CancellationToken cancellation
                     // so add a newline to the end of the query. See https://github.com/Microsoft/sqlopsstudio/issues/1424
                     dbCommand.CommandText += Environment.NewLine;
 
+                    EnsureConnectionIsOpen(conn);
```

Comment:
Your fix will address the particular symptom, I am not sure whether you are aware, but in ADO.NET, the connection state might appear open even though the underlining connection is broken until you try to use the connection to execute some commands, the state will be changed accordingly only when the command cannot be executed.

I would suggest to introduce a retry helper with retry strategy and use it whenever we need reliable connection.

Reply:
ReliableConnection is only for DbConnection and DbCommand operations as far as I could see.  It doesn't do anything for DataReaders.  

I'll look to see if there's a way to incorporate DataReaders into that component as well, though in general I'd agree that elaborate retry for all data operations seems to be looking at symptoms instead of root problems.  For example, when we create the DbCommand immediately prior to the DataReader the retry logic is already applied, then a few milliseconds later the connection is dropped.  

I haven't seen anywhere in any of our tools code an attempt to add retries for all interactions with SqlClient.  And this creates concerns as to how are 3rd-parties supposed to build applications using SqlClient if the expectation is that every API needs to be retried several times on failure.  I'm not yet convinced we're looking in the correct place here.